sdk/cosmos/microsoft-azure-cosmos/src/main/java/com/azure/data/cosmos/internal/changefeed/implementation/DocumentServiceLeaseStore.java

```diff
@@ -91,12 +91,10 @@ public DocumentServiceLeaseStore(
         containerDocument.id(lockId);
         BridgeInternal.setProperty(containerDocument, com.azure.data.cosmos.internal.Constants.Properties.TTL, Long.valueOf(lockExpirationTime.getSeconds()).intValue());
 
-        DocumentServiceLeaseStore self = this;
-
         return this.client.createItem(this.leaseCollectionLink, containerDocument, null, false)
             .map(documentResourceResponse -> {
                 if (documentResourceResponse.item() != null) {
-                    self.lockETag = documentResourceResponse.properties().etag();
+                    this.lockETag = documentResourceResponse.properties().etag();
```

Comment:
same story about thread safety here.

lockEtag will be set on a thread which subscribes to the returned Mono. I wonder if a second thread access lockEtag if it sees its most recent value?

Reply:
We're good here as well in terms of thread safety. A DocumentServiceLeaseStore instance is only used during the bootstrapping initialization which is single threaded for a given instance of ChangeFeedProcessor. Once the lock is acquired via the lease document (see the previous Bootstrapper comment) we are running in a critical section until the releaseInitializationLock() is called (which is the method that uses this.lockETag eventually).

Reply:
hmm! I don't think this is single threaded. I see 3 threads accessing lockEtag on DocumentServiceLeaseStore.java

1. `this.lockETag = documentResourceResponse.properties().etag();` is invoked by thread observing the result of Mono returned by `acquireInitializationLock(.)`

2. ` accessCondition.condition(this.lockETag)` is invoked by the thread calling `releaseInitializationLock(.)`

3. `this.lockETag = null;` is invoked by the thread observing the result of Mono returned by `releaseInitializationLock(.)` 

Let's talk on teams about this in more detail.

Reply:
the 3 threads in question run sequentially and not in parallel. Anyway, I've added "volatile" to the offending class field.